### PR TITLE
feat(components): Implement the new sorting behavior / alignment to DataTable component in Atlantis [JOB-58028]

### PR DESCRIPTION
**Pull request by kyleherrmann**
## Motivations

- New behaviour provides a better UX when sorting on `DataTable` columns

### Changed

- Updated sorting behaviour to use a new icon, SVG elements were added as a new component in the DataTable directory. [JOB-58028]

## Testing

- Run the Atlantis web app locally and navigate to the [sorting example](http://localhost:3333/components/data-table/#disable-sorting-of-a-specific-column), validate that the new icons behaves as expected and are hidden when sorting is disabled on a column

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

[JOB-58028]: https://jobber.atlassian.net/browse/JOB-58028?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/components/src/DataTable/Header.tsx [hunk 1]
```diff
@@ -46,14 +46,22 @@ export function Header<T extends object>({
...
-                    {flexRender(
-                      header.column.columnDef.header,
-                      header.getContext(),
-                    )}
+                    <>
+                      {flexRender(
+                        header.column.columnDef.header,
+                        header.getContext(),
+                      )}
+                      {header.column.getCanSort() ||
```
Comment (kyleherrmann) [inline packages/components/src/DataTable/Header.tsx:54]: Had to use a turnery for this logic to work.. not sure why I couldn't use a conditional `&&` - any insight is appreciated

**Comment by iremdesign** [comment]
At equilibrium, both the sort indicators are set to `#657884` . When you initiate the sorting behaviour, top icon is set to `#012939` and the bottom one is set to `#C1C9CE` . If you were to click on the sorting indicators again the colours for the top and bottom icon will alternate. On third click, when you decide to go back to equilibrium the colour should be set back to `#657884` .

In your demo, the sort indicator colour is set to `#C1C9CE` and I don't think I saw the `#657884` throughout the demo.

cc: @chris-at-jobber @5763048

File: packages/components/src/DataTable/DataTable.mdx [hunk 2]
```diff
@@ -332,6 +332,142 @@ import { DataTable } from "@jobber/components/DataTable";
...
+    pagination={{ manualPagination: false, itemsPerPage: [10, 20, 30] }}
+    sorting={{ manualSorting: false }}
+    onRowClick={row => alert(JSON.stringify(row.original, null, 2))}
+    columns={[
+      {
+        accessorKey: "name",
+        cell: info => info.getValue(),
+        header: "Name",
+        enableSorting: false,
+        enableMultiSort: false,
```
Comment (danlaurent) [inline packages/components/src/DataTable/DataTable.mdx:443]: Do we need both to disable sorting for a specific column? 🤔

File: packages/components/src/DataTable/SortIcon.css [hunk 3]
```diff
@@ -0,0 +1,17 @@
+.sortIcon {
+  margin: 0;
+  /* padding-top: calc(var(--space-smaller) + (var(--space-smaller) * 0.2));
+  padding-left: var(--space-small); */
```
Comment (danlaurent) [inline packages/components/src/DataTable/SortIcon.css:4]:
```suggestion
```

**Review by danlaurent — COMMENTED** [review]
Code LGTM, just left a cleanup suggestion.
I got a question regarding the sort icons alignment, is it still expected that for right aligned columns, to show the sort icons at the right end of the cell?

File: packages/components/src/DataTable/DataTable.mdx (hunk 2 above)
Comment (kyleherrmann) [reply packages/components/src/DataTable/DataTable.mdx:443]: Appears not, I'll remove it 🙂

**Comment by kyleherrmann** [comment]
> Code LGTM, just left a cleanup suggestion. I got a question regarding the sort icons alignment, is it still expected that for right aligned columns, to show the sort icons at the right end of the cell?

@iremdesign 👀

**Comment by kyleherrmann** [comment]
> Code LGTM, just left a cleanup suggestion. I got a question regarding the sort icons alignment, is it still expected that for right aligned columns, to show the sort icons at the right end of the cell?

Isn't alignment static? All columns are aligned on the right per designs

**Comment by kyleherrmann** [comment]
> Code LGTM, just left a cleanup suggestion. I got a question regarding the sort icons alignment, is it still expected that for right aligned columns, to show the sort icons at the right end of the cell?

Actually it appears to be a mix, but @iremdesign has signed off on the current implementation with consistent alignment and icons on the right.

Screenshot 2023-01-13 at 12 26 17 PM

**Comment by danlaurent** [comment]
> > Code LGTM, just left a cleanup suggestion. I got a question regarding the sort icons alignment, is it still expected that for right aligned columns, to show the sort icons at the right end of the cell?
> 
> Actually it appears to be a mix, but @iremdesign has signed off on the current implementation with consistent alignment and icons on the right.
> 
> Screenshot 2023-01-13 at 12 26 17 PM

But here for example, shouldn't the icons be right aligned with the end of the `points, change and power` values?
Screen Shot 2023-01-13 at 17 44 11

**Comment by danlaurent** [comment]
Like this example:
Screen Shot 2023-01-13 at 17 48 24

**Comment by iremdesign** [comment]
For text-based columns, we don't need any comparison so they can be aligned to column copy (i.e. Franchisee Account Name) 

For numerical columns (i.e. Job Number) that need no comparison, they can be aligned to column copy 

For numerical columns with monetary value (i.e. Total) the sort indicator will align to the very end of the values.

For numerical columns (such as Total # of Quotes), they can be aligned to column copy but we can also align to the every end of the values. I prefer the latter because the user can easily see compare which franchisee has the most amount of quotes etc.

@danlaurent There is something strange happening in your screenshot where the values are aligning to the end of the sort indicator container (24 x 24) and not the indicator icon. 

Expected alignment:

Screen Shot 2023-01-13 at 2 05 24 PM

**Comment by kyleherrmann** [comment]
> For text-based columns, we don't need any comparison so they can be aligned to column copy (i.e. Franchisee Account Name)
> 
> For numerical columns (i.e. Job Number) that need no comparison, they can be aligned to column copy
> 
> For numerical columns with monetary value (i.e. Total) the sort indicator will align to the very end of the values.
> 
> For numerical columns (such as Total # of Quotes), they can be aligned to column copy but we can also align to the every end of the values. I prefer the latter because the user can easily see compare which franchisee has the most amount of quotes etc.
> 
> @danlaurent There is something strange happening in your screenshot where the values are aligning to the end of the sort indicator container (24 x 24) and not the indicator icon.
> 
> Expected alignment:
> 
> Screen Shot 2023-01-13 at 2 05 24 PM

Is this AC a critical? I can adjust alignment based on data type but that will take quite some time. Can we align on the right hand side to the end of the column value for all data types?

Screenshot 2023-01-13 at 1 53 17 PM

**Comment by iremdesign** [comment]
The ticket also lists the design requirements so ideally we work on this as part of the ticket. This is an important functionality we'll use in all Franchise Reports so we'll have to work on this one way or the other. When you say quite some time, how many days are we looking at?

**Comment by kyleherrmann** [comment]
> The ticket also lists the design requirements so ideally we work on this as part of the ticket. This is an important functionality we'll use in all Franchise Reports so we'll have to work on this one way or the other. When you say quite some time, how many days are we looking at?

It seems the design requirements have a static alignment for the date icon 🤔 but if we just reference the child data and move the icon to the end of the cell when alignment is set to the right I can do this within a half day.

Screenshot 2023-01-13 at 3 02 03 PM

**Comment by iremdesign** [comment]
To clarify, I was referring to the fact that the sort indicator is not going to be a prefix but a suffix at all times. It will be added following the header column copy. 

Where it is aligned to will be determined by the column type. 

Numerical columns (comparison needed): the values in the column will be aligned right, to the end of the sorting indicator icon as seen here. 
Screen Shot 2023-01-13 at 4 24

Numerical columns (no comparison needed): the values in the column will be aligned left, the sorting indicator will come right after the text

Example: 

Screen Shot 2023-01-13 at 4 25 08 PM

Text-based columns (doesn't matter if the text is long or not): no comparison needed, the values will be left-aligned, the sorting indicator will come right after the text. 

Example: 
Screen Shot 2023-01-13 at 4 30 18 PM

**Comment by iremdesign** [comment]
We'll have a lot of numerical columns (monetary value) where the child content (values) are used for comparison so ideally we would work on this as part of this ticket. See examples here: 

Group 246

**Comment by kyleherrmann** [comment]
> To clarify, I was referring to the fact that the sort indicator is not going to be a prefix but a suffix at all times. It will be added following the header column copy.
> 
> Where it is aligned to will be determined by the column type.
> 
> Numerical columns (comparison needed): the values in the column will be aligned right, to the end of the sorting indicator icon as seen here. Screen Shot 2023-01-13 at 4 24
> 
> Numerical columns (no comparison needed): the values in the column will be aligned left, the sorting indicator will come right after the text
> 
> Example:
> 
> Screen Shot 2023-01-13 at 4 25 08 PM
> 
> Text-based columns (doesn't matter if the text is long or not): no comparison needed, the values will be left-aligned, the sorting indicator will come right after the text.
> 
> Example: Screen Shot 2023-01-13 at 4 30 18 PM

Yes pattern makes sense now 👍 This alignment is already set manually so we can reference that setting which is much less complicated.

**Comment by iremdesign** [comment]
Just FYI, if you look at the screenshot @danlaurent shared with us there is an issue where the values are aligning to the very end of the invisible container / frame that has the sort indicator icon. Could we please align to the end of the sorting indicator icon and not to the end of the invisible container / frame? Otherwise, there will be a strange space like you see in the yellow solid line and the dashed red line 

 
Group 247

**Comment by danlaurent** [comment]
> Just FYI, if you look at the screenshot @danlaurent shared with us there is an issue where the values are aligning to the very end of the invisible container / frame that has the sort indicator icon. Could we please align to the end of the sorting indicator icon and not to the end of the invisible container / frame? Otherwise, there will be a strange space like you see in the yellow solid line and the dashed red line

Yeah, it was just an example I quickly made to show which elements alignment I was talking about

**Comment by kyleherrmann** [comment]
> To clarify, I was referring to the fact that the sort indicator is not going to be a prefix but a suffix at all times. It will be added following the header column copy.
> 
> Where it is aligned to will be determined by the column type.
> 
> Numerical columns (comparison needed): the values in the column will be aligned right, to the end of the sorting indicator icon as seen here.
> 
> Numerical columns (no comparison needed): the values in the column will be aligned left, the sorting indicator will come right after the text
> 
> Example:
> 
> Text-based columns (doesn't matter if the text is long or not): no comparison needed, the values will be left-aligned, the sorting indicator will come right after the text.
> 
> Example:

AC is clear here but the alignment is set in the typography component... there is no indicator available for what the column alignment is set to from the data table component. We may need to redo the implementation of the icon to exist outside of the table which will likely take a day or two. Thoughts @danlaurent?

**Comment by kyleherrmann** [comment]
> > To clarify, I was referring to the fact that the sort indicator is not going to be a prefix but a suffix at all times. It will be added following the header column copy.
> > Where it is aligned to will be determined by the column type.
> > Numerical columns (comparison needed): the values in the column will be aligned right, to the end of the sorting indicator icon as seen here.
> > Numerical columns (no comparison needed): the values in the column will be aligned left, the sorting indicator will come right after the text
> > Example:
> > Text-based columns (doesn't matter if the text is long or not): no comparison needed, the values will be left-aligned, the sorting indicator will come right after the text.
> > Example:
> 
> AC is clear here but the alignment is set in the typography component... there is no indicator available for what the column alignment is set to from the data table component. We may need to redo the implementation of the icon to exist outside of the table which will likely take a day or two. Thoughts @danlaurent?

I suppose we could generate an array when we render the table that indicates which columns are aligned to the right using their index.

**Comment by danlaurent** [comment]
> > > To clarify, I was referring to the fact that the sort indicator is not going to be a prefix but a suffix at all times. It will be added following the header column copy.
> > > Where it is aligned to will be determined by the column type.
> > > Numerical columns (comparison needed): the values in the column will be aligned right, to the end of the sorting indicator icon as seen here.
> > > Numerical columns (no comparison needed): the values in the column will be aligned left, the sorting indicator will come right after the text
> > > Example:
> > > Text-based columns (doesn't matter if the text is long or not): no comparison needed, the values will be left-aligned, the sorting indicator will come right after the text.
> > > Example:
> > 
> > 
> > AC is clear here but the alignment is set in the typography component... there is no indicator available for what the column alignment is set to from the data table component. We may need to redo the implementation of the icon to exist outside of the table which will likely take a day or two. Thoughts @danlaurent?
> 
> I suppose we could generate an array when we render the table that indicates which columns are aligned to the right using their index.

@kyleherrmann  I thought of something simpler, let's talk in my office hours, does it work for you?

**Comment by kyleherrmann** [comment]
@danlaurent New implementation pushed

File: packages/components/src/DataTable/Header.tsx [hunk 4]
```diff
@@ -42,18 +42,26 @@ export function Header<T extends object>({
                   width: header.getSize(),
                   minWidth: header.column.columnDef.minSize,
                   maxWidth: header.column.columnDef.maxSize,
+                  paddingRight: 0,
```
Comment (danlaurent) [inline packages/components/src/DataTable/Header.tsx:45]: I believe this should be applied only if the column is sortable right? If so it's easier to add it into the `sortableColumn` css class.

**Review by danlaurent — CHANGES_REQUESTED** [review]
Just found a missing treatment for the header spacing / padding.

**Comment by iremdesign** [comment]
@danlaurent @kyleherrmann one remaining question i had is is the new behaviour (right alignment for numerical columns) going to get applied to all columns with numerical values? i.e. Job Titles (no comparison needed).

**Comment by kyleherrmann** [comment]
> @danlaurent @kyleherrmann one remaining question i had is is the new behaviour (right alignment for numerical columns) going to get applied to all columns with numerical values? i.e. Job Titles (no comparison needed).

It is set manually for each column so we can match [this pattern](https://github.com/GetJobber/atlantis/pull/972#issuecomment-1382584605)

**Review by danlaurent — APPROVED** [review]
LGTM! 🚀

**Pull request merged**